### PR TITLE
Add state spaces Sphere, Torus and Mobius.

### PR DESCRIPTION
**Pull request by aorthey**
Hello everyone,

to make sure that a planner works correctly, I found it super useful to have some simple geometrical spaces which we can visualize. Here are three which I often use: Sphere, Torus and Mobius strip. They all come with correct uniform sampling methods and correct interpolation/distance metrics attached (see pictures to get an idea). 

I thought someone else might be interested in using them, so I wanted to share them. 

Best,
Andreas

sphere_interpolation_thumbnailsphere_samples_thumbnailtorus_interpolation_thumbnailtorus_samples_thumbnailmobius_interpolation_thumbnailmobius_samples_thumbnail

### CODE REVIEW

File: src/ompl/base/spaces/src/MobiusStateSpace.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+State *MobiusStateSpace::allocState() const
+{
+    auto *state = new StateType();
+    allocStateComponents(state);
+    return state;
+}
+
+void MobiusStateSpace::freeState(State *state) const
```
Comment (mamoll) [inline src/ompl/base/spaces/src/MobiusStateSpace.cpp:143]: Do you need this method? The method doesn't override anything the base class would do.

File: src/ompl/base/spaces/src/SphereStateSpace.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+ompl::base::State *ompl::base::SphereStateSpace::allocState() const
+{
+    auto *state = new StateType();
+    allocStateComponents(state);
+    return state;
+}
+
+void ompl::base::SphereStateSpace::freeState(State *state) const
```
Comment (mamoll) [inline src/ompl/base/spaces/src/SphereStateSpace.cpp:129]: Not needed

File: src/ompl/base/spaces/src/TorusStateSpace.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+TorusStateSampler::TorusStateSampler(const StateSpace *space) : StateSampler(space)
+{
+}
+
+void TorusStateSampler::sampleUniform(State *state)
+{
+    // https://stackoverflow.com/questions/26300510/generating-random-points-on-a-surface-of-an-n-dimensional-torus
+    // Based on publication "Random selection of points distributed on curved surfaces."
+    // Link: https://iopscience.iop.org/article/10.1088/0031-9155/32/10/009/pdf
+    const TorusStateSpace *T = static_cast<const TorusStateSpace *>(space_);
```
Comment (mamoll) [inline src/ompl/base/spaces/src/TorusStateSpace.cpp:56]:
```suggestion
    const auto *T = static_cast<const TorusStateSpace *>(space_);
```

File: src/ompl/base/spaces/src/TorusStateSpace.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+State *TorusStateSpace::allocState() const
+{
+    auto *state = new StateType();
+    allocStateComponents(state);
+    return state;
+}
+
+void TorusStateSpace::freeState(State *state) const
```
Comment (mamoll) [inline src/ompl/base/spaces/src/TorusStateSpace.cpp:129]: Not needed

**Review by mamoll — CHANGES_REQUESTED** [review]
Looks good! I have left some minor comments. I hope to get to your older massive PR soon.

**Comment by aorthey** [comment]
Thank you Mark! Not sure why freeState was still in there. Should be good now.

**Comment by mamoll** [comment]
If it's not too much trouble, could you extend the [state space tests](https://github.com/ompl/ompl/blob/main/tests/base/state_spaces.cpp) with some basic tests for your new spaces?

**Comment by marlinstrub** [comment]
Thanks for sharing these @aorthey

**Comment by aorthey** [comment]
Ok I added the state space tests. I found two small bugs (Mobius used same header define as Torus, and the sphere distance function used a wrong angle). Now it looks good and tests run through.

I also added a function "toVector" which projects the parameterization of the spaces into 3D space (which would be useful if anyone wants to actually visualize it).

**Review by mamoll — APPROVED** [review]

**Pull request merged**